enhancements/cluster-scope-secret-volumes/shared-resource-validation.md

```diff
@@ -0,0 +1,413 @@
+
+As a developer or administrator using the Shared Resource CSI Driver I want necessary debug of the validation mechanism to be captured by
+OCP must-gather process.
+
+
+### API Extensions
+
+Quick note:  When discussing API, 'Shared Resource' on its own == SharedConfigMap / SharedSecret, and 'Reference' == the corresponding
+Secret or ConfigMap.
+
+A [ValidatingAdmissionWebhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook)
+will be used to inspect all created/modified Shared Resources, and make sure their names do not start
+with the reserved prefix "openshift-" unless their Shared Resource name and Reference key are found in a 
+well known list of OCP pre-populated Shared Resource.  That list will be in well known ConfigMaps that resides 
+in the managed CSI Driver namespace.  Components / operators other than Shared Resources that want to create pre-populated
+SharedSecrets or SharedConfigMaps should submit PRs against the Shared Resource CSI Driver Operator Repo to update the 
+corresponding ConfigMap with the necessary information.  This is akin to how we create in-payload image streams via the PRs against the 
+samples operator today.
+
+There will be a "ocp-sharedsecret-list" and "ocp-sharedconfigmap-list" ConfigMap.  The key of an entry is the Shared Resource
+name.  The value should be of the form "namespace:name" in order to capture. 
+
+The "openshift-*" naming restriction is new as compared to the initial tech preview release of Shared Resources in 4.10.
+
+A ValidatingAdmissionWebhook will be used to inspect all created Pods, and k8s Pod derived types (Deployments, DaemonSets, etc.), 
+as well Builds and BuildConfigs, and make sure no read-write Shared Resource CSI Volumes are specified.
+
+This restriction exists in 4.10, but is only enforced at the CSI driver level.
+
+The idea of validating the References of any future well-known, pre-populated Shared Resources
+(like the entitlement secret) was also raised during the review of the 4.10 blog post fof Shared Resources.  
+
+The discussion landed around these points:
+- A golang constant will declare the name of a k8s Condition that other components (i.e. OC Operators besides Shared Resources) will
+set on any Shared Resource that they pre-populate into OCP clusters.
+- REMINDER: the status structs of Shared Resources have an array of Conditions.  See 
+[here](https://github.com/openshift/api/blob/dbc82f2a3bc8b07e39628d36aee19d5a62847e2c/sharedresource/v1alpha1/types_shared_configmap.go#L83-L88) and
+[here](https://github.com/openshift/api/blob/dbc82f2a3bc8b07e39628d36aee19d5a62847e2c/sharedresource/v1alpha1/types_shared_secret.go#L81-L86)
+- The components who create pre-populated Shared Resources should have a controller that performs whatever validations
+they deem possible to verify the References, and those controllers are what will update the Status Condition arrays
+for the pre-populated Shared Resources
```

Comment:
It seems odd to me that a separate controller would need to update status. What status conditions will these controllers set?

Reply:
Yeah I need to do a better job of describing what @deads2k and I discussed.  I'll post a concrete example here (maybe I even put some form of that in the EP).  Assuming that concrete example helps clear things up for you @adambkaplan, I'll then move on to improving the EP itself.

So consider:
- the Insights Operator starts creating a SharedSecret to encapsulate the entitlement secret they maintain
- they seem to already being doing some "validations" around that secret 
- maybe then even have more in the works?
- maybe they agree to do some items that we or @deads2k suggest?
- if they discover something wrong with the entitlement secret, the insights operator updates the status of the SharedSecret they create for using it, setting true/false, reason, message on our well known "validity" condition, explaining someting is wrong
- Very similar to what CVO operators do with the well known conditions CVO defines for the ClusterOperator object
- then OpenShift Build development, we we get the case that says "entitled builds don't work", will ask them to check the conditions on the SharedSecret that Insights Operator creates
- the Reason/Message there helps the customer figure out what he needs to do to fix his entitlement credentials (maybe he has to log onto console.redhat.com and do something) and then his entitlement builds work

Does that help @adambkaplan ?  If it does, including perhaps you accepting the CVO operator analogy, do you think you are on board?

thanks

Reply:
Now that I see the full impact of my idea above, I'm less convinced adding something to `spec` is a good idea. Leaving `Ready` as an unmanaged condition that the CSI driver checks is perhaps the best option. We would want to document this in the status API.